### PR TITLE
added `-R --remove-duplicate-data` in `csv timeline`

### PR DESCRIPTION
**Pull request by hitenkoku**
## What Changed

- Added `-R, --remove-duplicate-data` option in `csv-timeline` command, for remove duplicate `Details` field (`%Details%`, `%AllFieldInfo%`, `%ExtraFieldInfo%`) output. 

I would appreciate it if you could review when you have time.

### CODE REVIEW

**Comment by hitenkoku** [comment]
## Evidence

- hayabusa-sample-evtx repository (137.1MB)

```
> ./1056.exe csv-timeline -d ..\hayabusa-sample-evtx\ -o 1056-r.csv --debug -R -C  -q
...
Total event log files: 584
Total file size: 137.1 MB

Loading detections rules. Please wait.

Excluded rules: 30
Noisy rules: 12 (Disabled)

Deprecated rules: 169 (4.54%) (Disabled)
Experimental rules: 2001 (53.78%)
Stable rules: 225 (6.05%)
Test rules: 1495 (40.18%)
Unsupported rules: 43 (1.16%) (Disabled)

Hayabusa rules: 152
Sigma rules: 3569
Total enabled detection rules: 3721

Output profile: standard

Scanning in progress. Please wait.

...

Results Summary:
First Timestamp: 2009-07-14 13:56:45.074 +09:00
Last Timestamp: 2023-01-24 20:54:42.899 +09:00

Events with hits / Total events: 19,604 / 47,472 (Data reduction: 27,868 events (58.70%))

Total | Unique detections: 32,062 | 629
Total | Unique critical detections: 56 (0.17%) | 22 (3.50%)
Total | Unique high detections: 6,042 (18.84%) | 279 (44.36%)
Total | Unique medium detections: 1,894 (5.91%) | 205 (32.59%)
Total | Unique low detections: 6,002 (18.72%) | 70 (11.13%)
Total | Unique informational detections: 18,068 (56.35%) | 53 (8.43%)

Dates with most total detections:
critical: 2019-07-19 (15), high: 2016-09-20 (3,652), medium: 2021-04-22 (199), low: 2016-09-20 (3,708), informational: 2016-08-19 (2,104)

Top 5 computers with most unique detections:
critical: MSEDGEWIN10 (8), IEWIN7 (3), FS03.offsec.lan (2), IE10Win7 (2), rootdc1.offsec.lan (2)
high: MSEDGEWIN10 (118), IEWIN7 (70), FS03.offsec.lan (30), fs03vuln.offsec.lan (29), IE10Win7 (24)
medium: MSEDGEWIN10 (77), IEWIN7 (48), FS03.offsec.lan (21), fs03vuln.offsec.lan (19), srvdefender01.offsec.lan (16)
low: MSEDGEWIN10 (32), FS03.offsec.lan (17), IEWIN7 (15), fs03vuln.offsec.lan (15), fs01.offsec.lan (11)
informational: IEWIN7 (18), MSEDGEWIN10 (18), fs01.offsec.lan (15), PC01.example.corp (14), FS03.offsec.lan (13)



Saved file: 1056-r.csv (25.5 MB)

Elapsed time: 00:00:07.640
Rule Parse Processing Time: 00:00:01.079
Analysis Processing Time: 00:00:06.147
Output Processing Time: 00:00:00.413

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:     2.0 GiB     2.0 GiB     0           2.0 GiB
 committed:   781.7 MiB     1.2 GiB   804.6 MiB   493.7 MiB
     reset:     0
    purged:   370.3 MiB
   touched:    64.2 KiB     1.1 MiB     3.5 GiB    -3.5 GiB                          ok
  segments:    16          19           8          11                                not all freed!
-abandoned:     0           0           0           0                                ok
   -cached:     0           0           0           0                                ok
     pages:     0           0          32.5 Ki    -32.5 Ki                           ok
-abandoned:     0           0           0           0                                ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     6.9 Ki
    resets:     0
    purges:   306
   threads:    32          32           0          32                                not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:     7.645 s
   process: user: 28.640 s, system: 0.718 s, faults: 258956, rss: 675.4 MiB, commit: 766.3 MiB
```

- all-evtx (6.1GB)

```
> ./1056.exe csv-timeline -d ..\all-evtx\ -o 1056-all-evtx-r.csv --debug -R -C -q
...
Total event log files: 1858
Total file size: 6.1 GB

Loading detections rules. Please wait.

Excluded rules: 30
Noisy rules: 12 (Disabled)

Deprecated rules: 169 (4.54%) (Disabled)
Experimental rules: 2001 (53.78%)
Stable rules: 225 (6.05%)
Test rules: 1495 (40.18%)
Unsupported rules: 43 (1.16%) (Disabled)

Hayabusa rules: 152
Sigma rules: 3569
Total enabled detection rules: 3721

Output profile: standard

Scanning in progress. Please wait.

...

Results Summary:
First Timestamp: 2009-07-14 13:56:45.074 +09:00
Last Timestamp: 2022-09-18 23:37:13.088 +09:00

Events with hits / Total events: 1,594,151 / 4,817,181 (Data reduction: 3,223,030 events (66.91%))

Total | Unique detections: 1,627,313 | 158
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 11,646 (0.72%) | 19 (12.03%)
Total | Unique medium detections: 11,017 (0.68%) | 44 (27.85%)
Total | Unique low detections: 1,054,590 (64.81%) | 45 (28.48%)
Total | Unique informational detections: 550,060 (33.80%) | 50 (31.65%)

Dates with most total detections:
critical: n/a, high: 2022-09-18 (4,438), medium: 2022-02-08 (4,802), low: 2022-09-18 (911,842), informational: 2022-03-02 (206,073)

Top 5 computers with most unique detections:
critical: n/a
high: evtx-PC (10), DESKTOP-A8CALR3 (7), DESKTOP-6D0DBMB (7), Agamemnon (6), WIN-FPV0DSIC9O6.sigma.fr (4)
medium: Agamemnon (28), DESKTOP-6D0DBMB (20), DESKTOP-A8CALR3 (19), evtx-PC (13), WIN-FPV0DSIC9O6 (9)
low: DESKTOP-6D0DBMB (29), DESKTOP-A8CALR3 (24), Agamemnon (23), evtx-PC (15), WIN-FPV0DSIC9O6 (13)
informational: DESKTOP-6D0DBMB (36), DESKTOP-A8CALR3 (35), WIN-TKC15D7KHUR (32), Agamemnon (31), WIN-FPV0DSIC9O6.sigma.fr (30)

╭───────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Top critical alerts:                        Top high alerts:                                      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ n/a                                         File Creation Date Changed to Another Yea... (10,490) │
│ n/a                                         Windows Shell/Scripting Application File ... (991)    │
│ n/a                                         Proc Exec (Non-Exe Filetype) (45)                     │
│ n/a                                         SysmonEnte Usage (33)                                 │
│ n/a                                         Disabling Windows Event Auditing (29)                 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ Top medium alerts:                          Top low alerts:                                       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ Raw Access Read (8,544)                     Proc Access (1,020,252)                               │
│ EVTX Created In Uncommon Location (986)     Possible Timestomping (31,784)                        │
│ Proc Injection (673)                        System Drawing DLL Load (1,030)                       │
│ Process Ran With High Privilege (191)       Creation of an Executable by an Executabl... (382)    │
│ Use Short Name Path in Command Line (133)   Firewall Rule Modified In The Windows Fir... (254)    │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ Top informational alerts:                                                                         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ File Created (422,037)                      Pipe Created (9,044)                                  │
│ File Deleted (53,696)                       Net Conn (8,755)                                      │
│ Proc Exec (18,780)                          DNS Query (5,108)                                     │
│ Pipe Conn (17,062)                          WMI Provider Started (681)                            │
│ Proc Terminated (12,388)                    WMI Modules Loaded (342)                              │
╰───────────────────────────────────────────╌───────────────────────────────────────────────────────╯

Saved file: 1056-all-evtx-r.csv (1.3 GB)

Elapsed time: 00:04:45.403
Rule Parse Processing Time: 00:00:01.751
Analysis Processing Time: 00:04:22.717
Output Processing Time: 00:00:20.933

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:     7.0 GiB     7.0 GiB     0           7.0 GiB
 committed:     5.7 GiB    31.2 GiB    26.1 GiB     5.0 GiB
     reset:     0
    purged:    12.7 GiB
   touched:    64.2 KiB     4.3 MiB    53.9 GiB   -53.9 GiB                          ok
  segments:    17          70          58          12                                not all freed!
-abandoned:     0           0           0           0                                ok
   -cached:     0           0           0           0                                ok
     pages:     0           0         727.4 Ki   -727.4 Ki                           ok
-abandoned:     0           0           0           0                                ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:   259.1 Ki
    resets:     0
    purges:    14.7 Ki
   threads:    32          32           0          32                                not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:   285.414 s
   process: user: 2460.796 s, system: 13.140 s, faults: 5715100, rss: 5.3 GiB, commit: 5.4 GiB
```

**Review by fukusuket — COMMENTED** [review]
@hitenkoku @YamatoSecurity  
I have a question about specfication!

If the following record is detected,
```
"Timestamp", ..., "RecordID","Details"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-02 00:00:00.000 +09:00", 2, "ABCDEF"
"2023-01-02 00:00:00.000 +09:00", 2, "ABCDEF"
"2023-01-02 00:00:00.000 +09:00", 2, "ABCDEF"
```

With the `-R` option, the current behavior is:(If `Details` are the same, remove duplicates)
```
"Timestamp", ..., "RecordID","Details"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
```

But I think the expected behavior is: ?(If **both** `Details` and `RecordID` are the same, remove duplicates)
```
"Timestamp", ..., "RecordID","Details"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "ABCDEF"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
```
I'm sorry if I misunderstood the specifications(or current behavior)...🙇
(If the current behavior is as expected, I'll test with the current spec!💪)

**Comment by fukusuket** [comment]
I misunderstood the specs. I'm sorry ... 🙇
The current behavior is:(If **both** `Details` and `Timestamp` are the same, remove duplicates ?)
```
"Timestamp", ..., "RecordID","Details"
"2023-01-01 00:00:00.000 +09:00", 1, "ABCDEF"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-01 00:00:00.000 +09:00", 1, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "ABCDEF"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
"2023-01-02 00:00:00.000 +09:00", 2, "DUP"
```
If the above recognition is correct, I think it's OK :) !!

**Comment by fukusuket** [comment]
I have one more question!

Run the following command.
`./hayabusa csv-timeline -d ../hayabusa-sample-evtx -o old.csv`
`./hayabusa csv-timeline -d ../hayabusa-sample-evtx -o new.csv -R`

then compare the results
```
fukusuke@fukusukenoAir hayabusa-2.6.0-all-platforms % head -n 3 old.csv
"Timestamp","Computer","Channel","EventID","Level","RecordID","RuleTitle","Details","ExtraFieldInfo"
"2013-10-24 01:16:13.843 +09:00","37L4247D28-05","Sec",4624,"info",2,"Logon (System) - Bootup","Type: 0 ¦ TgtUser: SYSTEM ¦ SrcComp: - ¦ SrcIP: - ¦ LID: 0x3e7","LogonGuid: 00000000-0000-0000-0000-000000000000 ¦ ProcessId: 0x4 ¦ ProcessName: ¦ SubjectLogonId: 0x0 ¦ SubjectUserSid: S-1-0-0 ¦ TargetDomainName: NT AUTHORITY ¦ TargetUserSid: S-1-5-18"
"2013-10-24 01:16:27.000 +09:00","37L4247D28-05","Sys",6005,"info",4,"Event Log Svc Started","-","Binary: DD070A0003001700100010001B00D8030000000000000000"
fukusuke@fukusukenoAir hayabusa-2.6.0-all-platforms % head -n 3 new.csv
"Timestamp","Computer","Channel","EventID","Level","RecordID","RuleTitle","Details","ExtraFieldInfo"
"2013-10-24 01:16:13.843 +09:00","37L4247D28-05","Sec",4624,"info",2,"Logon (System) - Bootup","Type: 0 ¦ TgtUser: SYSTEM ¦ SrcComp: - ¦ SrcIP: - ¦ LID: 0x3e7","LogonGuid: 00000000-0000-0000-0000-000000000000 ¦ ProcessId: 0x4 ¦ ProcessName: ¦ SubjectLogonId: 0x0 ¦ SubjectUserSid: S-1-0-0 ¦ TargetDomainName: NT AUTHORITY ¦ TargetUserSid: S-1-5-18"
"2013-10-24 01:16:27.000 +09:00","37L4247D28-05","Sys",6005,"info",4,"Event Log Svc Started","DUP","Binary: DD070A0003001700100010001B00D8030000000000000000"
```
Why is the second record replaced with `DUP`? 🤔

**Comment by hitenkoku** [comment]
@fukusuket I am sorry. It was due to the fact that my implementation had the wrong timing for initializing the variables.
I have corrected it and would greatly appreciate it if you could check it when you have a moment.

**Comment by YamatoSecurity** [comment]
@hitenkoku Thanks so much for the PR! @fukusuket Thanks for checking!
I ran some tests comparing the main 2.7.0-dev branch and this PR with and without the `-R` option.

This PR does a good job at reducing filesize of about 10% (10GB to 9GB), however, it seems to miss events but detects one more unique detection.

|                   | Main 2.7.0-dev | This PR without -R | This PR with -R |
|-------------------|----------------|--------------------|-----------------|
| Elasped Time      | 02:26:16.238   | 02:27:08.033       | 02:27:01.602    |
| Memory Usage      | 65.4 GB        | 49.4 GB            | 49.4 GB         |
| File Size         | 13.9 GB        | 10.0 GB            | 9.0 GB          |
| Total Detections  | 25,142,009     | 18,728,720         | 18,728,720      |
| Unique Detections | 370            | 371                | 371             |

Details:
```
Main 2.7.0-dev branch:
.\hayabusa-2.7.0-dev-main.exe csv-timeline -d D:\logs\608-eventlogs\ -u -U -D  -o main.csv --debug

First Timestamp: 2010-11-21 03:58:31.093 +00:00
Last Timestamp: 2020-02-25 17:17:00.835 +00:00

Events with hits / Total events: 22,749,640 / 73,049,133 (Data reduction: 50,299,493 events (68.86%))

Total | Unique detections: 25,142,009 | 370
Total | Unique critical detections: 95 (0.00%) | 7 (1.89%)
Total | Unique high detections: 86,152 (0.34%) | 95 (25.68%)
Total | Unique medium detections: 3,173,286 (12.62%) | 132 (35.68%)
Total | Unique low detections: 1,418,409 (5.64%) | 74 (20.00%)
Total | Unique informational detections: 20,464,067 (81.39%) | 62 (16.76%)

Saved file: main.csv (13.9 GB)

Elapsed time: 02:26:16.238
Errors were generated. Please check ./logs/errorlog-20230623_050444.log for details.

Rule Parse Processing Time: 00:00:02.369
Analysis Processing Time: 02:18:48.753
Output Processing Time: 00:07:25.071

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:    72.0 GiB    72.0 GiB     0          72.0 GiB
 committed:    65.5 GiB   339.3 GiB   274.7 GiB    64.6 GiB
     reset:     1.0 GiB
    purged:   152.9 GiB
   touched:    64.2 KiB    35.8 MiB   756.9 GiB  -756.9 GiB                          ok
  segments:    44         573         560          13                                not all freed!
-abandoned:     0           0           0           0                                ok
   -cached:     0           0           0           0                                ok
     pages:     0           0          10.5 Mi    -10.5 Mi                           ok
-abandoned:     0           0           0           0                                ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     2.5 Mi
    resets:    35
    purges:   304.9 Ki
   threads:    16          16           0          16                                not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:  8776.247 s
   process: user: 59178.203 s, system: 469.656 s, faults: 68466755, rss: 56.0 GiB, commit: 65.4 GiB
```

This PR without `-R`:
```
.\hayabusa-remove-dups.exe csv-timeline -d D:\logs\608-eventlogs\ -u -U -D  -o no-dups.csv --debug

First Timestamp: 2010-11-21 03:58:31.093 +00:00
Last Timestamp: 2020-02-25 17:17:00.835 +00:00

Events with hits / Total events: 16,352,779 / 73,049,133 (Data reduction: 56,696,354 events (77.61%))

Total | Unique detections: 18,728,720 | 371
Total | Unique critical detections: 95 (0.00%) | 7 (1.89%)
Total | Unique high detections: 86,279 (0.46%) | 95 (25.61%)
Total | Unique medium detections: 3,173,358 (16.94%) | 134 (36.12%)
Total | Unique low detections: 1,418,347 (7.57%) | 73 (19.68%)
Total | Unique informational detections: 14,050,641 (75.02%) | 62 (16.71%)

Saved file: no-dups.csv (10.0 GB)

Elapsed time: 02:27:08.033
Errors were generated. Please check ./logs/errorlog-20230622_224055.log for details.

Rule Parse Processing Time: 00:00:05.018
Analysis Processing Time: 02:22:54.612
Output Processing Time: 00:04:08.169

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:    52.0 GiB    52.0 GiB     0          52.0 GiB
 committed:    98.0 GiB   570.4 GiB   522.2 GiB    48.1 GiB
     reset:     1.0 GiB
    purged:   300.7 GiB
   touched:    64.2 KiB   270.8 MiB   757.4 GiB  -757.1 GiB                          ok
  segments:    32           4.2 Ki      4.2 Ki     12                                not all freed!
-abandoned:     0           0           0           0                                ok
   -cached:     0           0           0           0                                ok
     pages:     0           0          10.5 Mi    -10.5 Mi                           ok
-abandoned:     0           0           0           0                                ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     3.7 Mi
    resets:    40
    purges:   296.7 Ki
   threads:    16          16           0          16                                not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:  8830.956 s
   process: user: 60674.171 s, system: 439.875 s, faults: 94348959, rss: 48.3 GiB, commit: 49.4 GiB
```

This PR with `-R`:
```
.\hayabusa-remove-dups.exe csv-timeline -d D:\logs\608-eventlogs\ -u -U -D -R -o dups.csv --debug

Results Summary:
First Timestamp: 2010-11-21 03:58:31.093 +00:00
Last Timestamp: 2020-02-25 17:17:00.835 +00:00

Events with hits / Total events: 16,352,779 / 73,049,133 (Data reduction: 56,696,354 events (77.61%))

Total | Unique detections: 18,728,720 | 371
Total | Unique critical detections: 95 (0.00%) | 7 (1.89%)
Total | Unique high detections: 86,279 (0.46%) | 95 (25.61%)
Total | Unique medium detections: 3,173,358 (16.94%) | 134 (36.12%)
Total | Unique low detections: 1,418,347 (7.57%) | 73 (19.68%)
Total | Unique informational detections: 14,050,641 (75.02%) | 62 (16.71%)

Saved file: dups.csv (9.0 GB)

Elapsed time: 02:27:01.602
Errors were generated. Please check ./logs/errorlog-20230622_185156.log for details.

Rule Parse Processing Time: 00:00:04.072
Analysis Processing Time: 02:22:59.460
Output Processing Time: 00:03:58.056

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:    52.0 GiB    52.0 GiB     0          52.0 GiB
 committed:   100.9 GiB   575.2 GiB   524.4 GiB    50.8 GiB
     reset:     1.1 GiB
    purged:   303.2 GiB
   touched:    64.2 KiB   270.5 MiB   757.3 GiB  -757.1 GiB                          ok
  segments:    32           4.2 Ki      4.2 Ki     12                                not all freed!
-abandoned:     0           0           0           0                                ok
   -cached:     0           0           0           0                                ok
     pages:     0           0          10.5 Mi    -10.5 Mi                           ok
-abandoned:     0           0           0           0                                ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     3.8 Mi
    resets:    39
    purges:   297.1 Ki
   threads:    16          16           0          16                                not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:  8821.611 s
   process: user: 60742.281 s, system: 427.531 s, faults: 94995342, rss: 48.3 GiB, commit: 49.4 GiB
```


Besides missing events, there is some difference in the fields:
Main:
`"2018-08-11 15:47:15.929 +00:00","xxx","Sec",4634,"info",10977375,"Logoff","User: HealthMailbox250cf44 ¦ LID: 0x1021972ec","LogonType: 3 ¦ TargetDomainName: shieldbase ¦ TargetUserSid: S-1-5-21-3445421715-2530590580-3149308974-1163"`

This PR:
`"2018-08-11 15:47:15.929 +00:00","xxx","Sec",4634,"info",10977375,"Logoff","User: HealthMailbox250cf44 ¦ LID: 0x1021972ec ¦ Type: 3","TargetDomainName: shieldbase ¦ TargetUserSid: S-1-5-21-3445421715-2530590580-3149308974-1163"`

However, the %Details% for EID 4634 is `'User: %TargetUserName% ¦ LID: %TargetLogonId% ¦ Type: %LogonType%'` so this PR seems correct using the `Type` field and the main branch seems incorrect by using the `LogonType` field and putting it in the `%ExtraField%` column. This PR puts it in the right `%Details%` column.

I will take some more benchmarks with the updated PR.

File: src/afterfact.rs [hunk 1]
```diff
@@ -319,13 +323,15 @@ fn emit_csv<W: std::io::Write>(
...
     let mut author_list_cache: HashMap<CompactString, Nested<String>> = HashMap::new();
+    let mut prev_message: HashMap<CompactString, Profile> = HashMap::new();
     for (message_idx, time) in MESSAGEKEYS
         .lock()
         .unwrap()
         .iter()
         .sorted_unstable()
         .enumerate()
     {
+        // remove duplicate dataのための前レコード分の情報を保持する変数
```
Comment (fukusuket) [inline src/afterfact.rs:334]: Could you move the comment as well?

**Review by fukusuket — CHANGES_REQUESTED** [review]
It would be nice to move the comments as well :)

**Comment by YamatoSecurity** [comment]
@fukusuket Thanks for checking!
Maybe there is something weird with my Windows environment...
I checked the various versions of Hayabusa on my mac and all of them converted the `LogonType` to `Type`. I will try to do some more testing. I am using a dataset of 60GB so maybe it's too much data? Not sure..

**Comment by YamatoSecurity** [comment]
@fukusuket I just checked the updated PR (without `-R`) and now everything is the same as main, so it seems that the problem has been fixed.

**Comment by YamatoSecurity** [comment]
@fukusuket I now realize why there was a difference in unique detections and field names. when I changed branches in visual studio it used a different version of the rules... please ignore what I wrote about that. Everything looks good.

File: src/afterfact.rs (hunk 1 above)
Comment (hitenkoku) [reply src/afterfact.rs]: @fukusuket Thanks for your comment.
I adjusted comment in https://github.com/Yamato-Security/hayabusa/pull/1104/commits/e28d221e2d93a3cfad09f888f0a279c7a15af41e . 
I am very sorry for the repeated requests.
Please could you confirm this?

**Review by fukusuket — APPROVED** [review]
Thank you for fix :)  LGTM!!🚀

**Review by YamatoSecurity — APPROVED** [review]
@hitenkoku I rechecked things after the fix and everything looks great!
The output time, memory usage and file size is the same as before and now we can get almost a 15% decrease in file size in my tests!
14 GB -> 12 GB.

**Pull request merged**